server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java

```diff
@@ -155,11 +160,60 @@ public DocValueBits docsWithValue(LeafReaderContext context) throws IOException
                 return org.elasticsearch.index.fielddata.FieldData.docsWithValue(ordinals);
             }
 
-            public abstract SortedSetDocValues ordinalsValues(LeafReaderContext context)
-                    throws IOException;
+            /**
+             * Get a view into the leaf's ordinals and their {@link BytesRef} values.
+             * <p>
+             * Use {@link SortedSetDocValues#advanceExact}, {@link SortedSetDocValues#getValueCount},
+             * and {@link SortedSetDocValues#nextOrd} to fetch the ordinals. Use
+             * {@link SortedSetDocValues#lookupOrd} to convert form the ordinal
+             * number into the {@link BytesRef} value. Make sure to
+             * {@link BytesRef#deepCopyOf(BytesRef) copy} the result if you need
+             * to keep it.
+             * <p>
+             * Each leaf may have a different ordinal for the same byte array.
+             * Imagine, for example, an index where one leaf has the values
+             * {@code "a", "b", "d"} and another leaf has the values
+             * {@code "b", "c", "d"}. {@code "a"} has the ordinal {@code 0} in
+             * the first leaf and doesn't exist in the second leaf.
+             * {@code "b"} has the ordinal {@code 1} in the first leaf
+             * and {@code 0} in the second leaf. {@code "c"} doesn't exist in
+             * the first leaf and has the ordinal {@code 1} in the second leaf.
+             * And {@code "d"} gets the ordinal {@code 2} in both leaves.
+             * <p>
+             * This can be a huge pain if you have to compare the ordinals
```

Comment:
I think it might be better to replace `This can be a huge pain` with something like `You should proceed with caution`. Same meaning, but it changes a message of frustration for a message of care, so I think it might have a more positive effect on a future reader.